server/services/odysee.ts

```diff
@@ -0,0 +1,794 @@
+): string {
+	return nodeResponseUrl ?? locationHeader ?? fallback;
+}
+
+type M3U8Attrs = { BANDWIDTH?: unknown; RESOLUTION?: { height?: unknown }; URI?: unknown };
+type M3U8Playlist = { uri?: unknown; attributes?: M3U8Attrs };
+function isPlaylist(x: unknown): x is M3U8Playlist {
+	return !!x && typeof x === "object";
+}
+function toNum(x: unknown): number {
+	return typeof x === "number" ? x : 0;
+}
+function toStr(x: unknown): string | undefined {
+	return typeof x === "string" ? x : undefined;
+}
+
+// Parse M3U8 master and pick the highest-quality variant
+async function pickBestHlsVariant(masterUrl: string): Promise<string> {
+	try {
+		const v = await verifyStream(masterUrl);
+		if (!v.ok) {
+			log.debug?.("pickBestHlsVariant: master not verifiable", {
+				masterUrl,
+				status: v.status,
+			});
+			return masterUrl;
+		}
+		const r = await httpc.get(masterUrl, {
+			responseType: "text",
+			headers: {
+				Accept: "application/vnd.apple.mpegurl,application/x-mpegurl,*/*",
+			},
+			validateStatus: s => s >= 200 && s < 400,
+		});
+
+		const parser = new M3U8Parser();
+		parser.push(r.data);
+		parser.end();
+
+		const manifest = parser.manifest;
+		const raw = Array.isArray(manifest?.playlists) ? manifest.playlists : [];
+		const playlists = raw.filter(isPlaylist);
+		if (!playlists.length) {
+			return masterUrl;
+		}
+
+		playlists.sort(
+			(a, b) =>
+				toNum(b.attributes?.BANDWIDTH) - toNum(a.attributes?.BANDWIDTH) ||
+				toNum(b.attributes?.RESOLUTION?.height) - toNum(a.attributes?.RESOLUTION?.height)
+		);
+
+		const best = playlists[0];
+		const bestUri = toStr(best?.uri) ?? toStr(best?.attributes?.URI);
+		if (bestUri) {
+			const abs = resolveRelativeUrl(masterUrl, bestUri);
+			log.debug?.(`HLS: selected best variant ${abs}`);
+			return abs;
+		}
+	} catch (e: unknown) {
+		const msg = e instanceof Error ? e.message : String(e);
+		log.debug?.("pickBestHlsVariant failed", { masterUrl, err: msg });
+	}
+	return masterUrl;
+}
+
+// Verify stream via HEAD (fallback GET range), get final URL and MIME
+async function verifyStream(
+	url: string
+): Promise<{ ok: boolean; status?: number; mime?: string; finalUrl?: string }> {
+	try {
+		const head = await httpc.head(url, {
+			validateStatus: s => (s >= 200 && s < 300) || s === 405,
+			headers: { Accept: "application/json,text/plain,*/*" },
+		});
+		const mime = head.headers?.["content-type"];
+		const loc = head.headers?.["location"];
+		const finalUrl = responseFinalUrl(
+			head.request?.res?.responseUrl,
+			typeof loc === "string" ? loc : undefined,
+			url
+		);
+
+		return {
+			ok: head.status >= 200 && head.status < 300,
+			status: head.status,
+			mime: typeof mime === "string" ? mime : undefined,
+			finalUrl,
+		};
+	} catch {
+		try {
+			const get = await httpc.get(url, {
+				responseType: "arraybuffer",
+				validateStatus: s => s === 206 || s === 200,
+				headers: {
+					Range: "bytes=0-1",
+					Accept: "application/json,text/plain,*/*",
```

Comment:
There needs to be an explanation for why this callback exists. Does the first sometimes not work for certain videos? Does it sometimes not give the needed info? etc.